src/prefect/concurrency/__init__.py

```diff
@@ -0,0 +1,223 @@
+    mode: Union[Literal["concurrency"], Literal["rate_limit"]] = "concurrency",
+) -> List[MinimalConcurrencyLimitResponse]:
+    async with get_client() as client:
+        response = await wait_for_successful_response(
+            client.increment_concurrency_slots,
+            names=names,
+            slots=slots,
+            mode=mode,
+            retryable_status_codes=[423],
+        )
+        return _response_to_minimal_concurrency_limit_response(response)
+
+
+async def release_concurrency_slots(
+    names: List[str], slots: int, occupancy_seconds: float
+) -> List[MinimalConcurrencyLimitResponse]:
+    async with get_client() as client:
+        response = await client.release_concurrency_slots(
+            names=names, slots=slots, occupancy_seconds=occupancy_seconds
+        )
+        return _response_to_minimal_concurrency_limit_response(response)
+
+
+@asynccontextmanager
+async def aconcurrency(names: Union[str, List[str]], occupy: int = 1):
+    names = names if isinstance(names, list) else [names]
+
+    limits = await acquire_concurrency_slots(names, occupy)
+    acquisition_time = pendulum.now("UTC")
+    emitted_events = _emit_concurrency_acquisition_events(limits, occupy)
+
+    try:
+        yield
+    finally:
+        occupancy_seconds: float = (
+            pendulum.now("UTC") - acquisition_time
+        ).total_seconds()
+        await release_concurrency_slots(names, occupy, occupancy_seconds)
+        _emit_concurrency_release_events(limits, occupy, emitted_events)
+
+
+@contextmanager
+def concurrency(names: Union[str, List[str]], occupy: int = 1):
+    names = names if isinstance(names, list) else [names]
+
+    limits: List[MinimalConcurrencyLimitResponse] = _call_async_function_from_sync(
+        acquire_concurrency_slots, names, occupy
+    )
+    acquisition_time = pendulum.now("UTC")
+    emitted_events = _emit_concurrency_acquisition_events(limits, occupy)
+
+    try:
+        yield
+    finally:
+        occupancy_seconds: float = (
+            pendulum.now("UTC") - acquisition_time
+        ).total_seconds()
+        _call_async_function_from_sync(
+            release_concurrency_slots, names, occupy, occupancy_seconds
+        )
+        _emit_concurrency_release_events(limits, occupy, emitted_events)
+
+
+async def arate_limit(names: Union[str, List[str]], occupy: int = 1):
+    """Block execution until an `occupy` number of slots of the concurrency
+    limits given in `names` are acquired. Requires that all given concurrency
+    limits have a slot decay.
+
+    Args:
+        names: The names of the concurrency limits to acquire slots from.
+        occupy: The number of slots to acquire and holf from each limit.
+    """
+    names = names if isinstance(names, list) else [names]
+    limits = await acquire_concurrency_slots(names, occupy, mode="rate_limit")
+    _emit_concurrency_acquisition_events(limits, occupy)
+
+
+def rate_limit(names: Union[str, List[str]], occupy: int = 1):
```

Comment:
But now I see where your heartburn is: the rate limit implementation is not a context manager, so you still have a similar problem with the naming here. :(